serialize2

laxels opened the pull request:
A rewrite of `serialize.ts` and `serialize.py` which replaces every single value with a reference. Values are hashed by value (lol), so values which are identical (including arrays and objects) are always deduplicated when serialized and always share the same reference when deserialized.

Unserialized:
```
{
  num: 1,
  str: "asdf",
  arr: [1, "asdf"],
  obj: {
    a: 1,
    b: "asdf"
  }
}
```

Serialized:
```
{
  nodes: [
    1,
    "asdf",
    [0, 1],
    {a: 0, b: 1},
    {
      num: 0,
      str: 1,
      arr: 2,
      obj: 3
    }
  ],
  targetNodes: [4]
}
```

laxels commented on weave/weave_types.py:
```diff
@@ -144,6 +147,16 @@ def type_from_dict(d: typing.Union[str, dict]) -> "Type":
 
+    @staticmethod
+    def type_from_dict_use_cache(d: typing.Union[str, dict]) -> "Type":
+        cache_key = id(d)
```
I chose to use the `id()` of the dict as the cache key because `serialize2` ensures that objects with the same value share the same object reference.

However, if we want to use this in other places, we may want to do some value-based hash.

tssweeney replied:
`id` can be a foot gun in python as the id is allowed to be re-used after gcing the object.

laxels replied:
Good catch! Thanks for lmk.

Should be addressed via the context block you suggested.

laxels commented on weave-js/src/core/server/remoteHttp.ts:
```diff
@@ -252,7 +252,7 @@ export class RemoteHttpServer implements Server {
     const nodes = nodeEntries.map(e => e.node);
     const [payloads, originalIndexes] = this.opts.contiguousBatchesOnly
       ? serializeMulti(nodes)
```
I checked and `contiguousBatchesOnly` seems to always be `false`, so I did not convert `serializeMulti`. LMK if I should make a `serialize2Multi`.

tssweeney replied:
perhaps add this as a code comment?

laxels replied:
Added.

tssweeney commented on weave-js/src/core/model/graph/index.ts:
```diff
@@ -2,6 +2,7 @@ export * from './construction';
 export * from './editing';
 export * from './norm';
 export * from './serialize';
+export * from './serialize2';
```
nit: can we avoid `thing2` naming - maybe something that indicates the difference would be more helpful.

nit2: i think `serialize` and `serialize2` have a bunch of the same symbols. This import pattern (importing both) will end up with `serialize2`s + anything from `serialize` not overwrote by `serialize2`. Is this intended?

laxels replied:
Changed to `serializeAllValues`.

TypeScript complains when there are duplicate symbols exported, so there's no concept of "overwriting". Everything exported by `serialize2` (now `serializeAllValues`) had different names.

tssweeney commented on weave-js/src/core/model/graph/serialize2.ts:
```diff
@@ -0,0 +1,121 @@
+// Handle serialization/deserialization of CG, for remote execution
```
Can you add a header comment as to what is different here vs serialize?

laxels replied:
Added.

tssweeney commented on weave/serialize2.py:
```diff
@@ -0,0 +1,233 @@
+from typing import Any, Dict, Tuple, Union, TypedDict, MutableMapping
```
Same here: would be helpful to have comment in the header about how this differs from serialize

laxels replied:
Dunzo

laxels commented on weave/weave_types.py:
```diff
@@ -94,6 +94,9 @@ def type_name_to_type(type_name):
     return mapping.get(js_to_py_typename(type_name))
 
 
+type_from_dict_cache: dict[int, "Type"] = {}
```
Good suggest. Done.

tssweeney commented on weave-js/src/core/model/graph/serializeAllValues.ts:
```diff
@@ -0,0 +1,155 @@
+    // This does not properly stringify nested objects, but that's ok because
+    // the object values should already be replaced with serialized refs.
+    return JSON.stringify(v, Object.keys(v).sort());
+  }
```
i would not normally expect so see a `set`, but in theory could be possible. We might want to stringify the elements, then sort the stringified elements, then stringify the set of strings.

laxels replied:
Are you saying that we have JS `Set`s in the graph to serialize? Interesting. How does the old `serialize` code handle that? It must transform it into something else because `JSON.stringify`ing a set always just yields `"{}"`.

tssweeney reviewed (commented):
Last comment before approval: would be nice to have a parameterized unit test that asserts the two-way serialization/deserialization methods return the same object

laxels commented:
@tssweeney I actually implemented quite a few changes since your last viewing. Here is a summary:
- Added parametrized unit test as you requested. Parameters were taken from `serialize.test.ts`.
- Added proper handling of `Date` objects (I neglected to convert them to `{type: 'date', val: ...}` before) along with a test that verifies said proper handling.

I also added handling of `Set`s and `Map`s to the `hash` function per [your comment](https://github.com/wandb/weave/pull/107#discussion_r1253607020). However, I realized soon after that once we properly support those values (see [my comment](https://github.com/wandb/weave/pull/107#discussion_r1253611618) on why the current way we send those values to the server just doesn't work), the special handling of `Set`s and `Map`s in `hash` won't be necessary.

I've added a ton of comments with lots of TODOs around supporting `Set` and `Map` values the same way we're supporting `Date` values.

I do have one open question (maybe I should ask @nickpenaranda instead?). In the old `serialize.ts`, I noticed that [we're including a hashed typed node ID for output nodes](https://github.com/wandb/weave/blob/38df292ea9dd75aee5a6443cea6c79a5b2053156/weave-js/src/core/model/graph/serialize.ts#L305) in the serialized result. Do we need to include the `id` in the serialized output? I checked the Python deserialization code and it doesn't seem to be used, opting for [its own implementation of `node_id`](https://github.com/wandb/weave/blob/38df292ea9dd75aee5a6443cea6c79a5b2053156/weave/serialize.py#L96) instead. So I'm opting to leave it out for now, but wanted to confirm.

tssweeney commented:
> @tssweeney I actually implemented quite a few changes since your last viewing. Here is a summary:
> 
> * Added parametrized unit test as you requested. Parameters were taken from `serialize.test.ts`.
> * Added proper handling of `Date` objects (I neglected to convert them to `{type: 'date', val: ...}` before) along with a test that verifies said proper handling.
> 
> I also added handling of `Set`s and `Map`s to the `hash` function per [your comment](https://github.com/wandb/weave/pull/107#discussion_r1253607020). However, I realized soon after that once we properly support those values (see [my comment](https://github.com/wandb/weave/pull/107#discussion_r1253611618) on why the current way we send those values to the server just doesn't work), the special handling of `Set`s and `Map`s in `hash` won't be necessary.
> 
> I've added a ton of comments with lots of TODOs around supporting `Set` and `Map` values the same way we're supporting `Date` values.
> 
> I do have one open question (maybe I should ask @nickpenaranda instead?). In the old `serialize.ts`, I noticed that [we're including a hashed typed node ID for output nodes](https://github.com/wandb/weave/blob/38df292ea9dd75aee5a6443cea6c79a5b2053156/weave-js/src/core/model/graph/serialize.ts#L305) in the serialized result. Do we need to include the `id` in the serialized output? I checked the Python deserialization code and it doesn't seem to be used, opting for [its own implementation of `node_id`](https://github.com/wandb/weave/blob/38df292ea9dd75aee5a6443cea6c79a5b2053156/weave/serialize.py#L96) instead. So I'm opting to leave it out for now, but wanted to confirm.

Thanks for the updated - rereviewing now. I would need to resplunk this code to have an answer for node_id. Nick might know off hand

tssweeney commented on weave-js/src/core/model/graph/serializeAllValues.test.ts:
```diff
@@ -0,0 +1,122 @@
+        }),
+      }),
+      // `new Date(0)` should be serialized as `{type: 'date', val: '1970-01-01T00:00:00.000Z'}`
+      rhs: constDate(new Date(0)),
```
should there be a `opDateToNumber` on this side?

tssweeney reviewed (approved):
Looks good!

shawnlewis commented:
We don't need the ID in the request anymore, that was for WeaveJS server
